_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+_relationship_ between the available operations, rather than
+individual descriptions of each operation in isolation.
+
+* There's more than one way to implement a particular bit of
+functionality and, due to the tradeoffs between the available
+approaches, you don't want to choose just one. Perhaps you'll use
+one in some circumstances and another in other circumstances.
+
+In this situation, it is particularly valuable to clearly
+distinguish between the specification and implementation of a
+program, and to restrict proofs about the implementation to those
+that guarantee agreement with the specification, performing all
+other reasoning on the specification. The rest of this post will
+explore a simple library that has both of the attributes listed
+above: a specification and implementation of the well-known stack
+data type.
+
+The (arguably) most concise and natural specification of a stack
+comes from a description of a relationship between the push and pop
+operations. Namely, if you push an item onto a stack, and then pop
+an item off of the resulting stack, you should get back the item you
+originally pushed. Similarly, the value of the stack itself after
+these two operations should be unchanged: when you add something to
+the top and immediately remove it, the rest of the stack stays the
+same.
+
+On the implementation side, there are multiple ways to build a
+stack. Common implementation techniques include custom linked lists,
+consecutive arrays (reallocated when full), and the re-use of
+built-in data types with appropriate properties, such as sequences
+or vectors. Perhaps you wouldn't use more than one of these in a
+given application, but we can still look at how to straightforwardly
+show that multiple implementations, using different techniques, all
+fit the same notion of what it means to be a stack.
+
+To start with, let's specify what it means to be a stack in the
+following module.
+
+{% highlight dafny %}
+module StackSpecification {
+  // Most clients, including clients of implementations, get no details.
+  // This means that _all_ they know about stacks is what the two lemmas
+  // describe.
+  export
+    // The operations on a stack.
+    provides StackModel, PushModel, PopModel, EmptyModel, IsEmptyModel
+    // The properties of those operations.
+    provides PushPopVal, PushPopStack
+
+  // Implementation modules can access the definition of the model
+  // operations, to allow proofs of equivalence, by importing the
+  // `Private export set.
+  export Private
+    reveals StackModel, PushModel, PopModel, EmptyModel, IsEmptyModel
+    provides PushPopVal, PushPopStack
+
+  // We provide a concrete type to model a stack for simplicity of
+  // proof, since Dafny knows how to prove things about sequences. Note,
+  // however, that most clients don't get to see what this type is.
+  type StackModel<T> = seq<T>
+
+  // Check to see whether a stack is empty.
+  ghost function IsEmptyModel<T>(stk: StackModel<T>) : bool
```

Comment:
Why ghost?

Reply:
It's used on `StackModel` values rather than stack implementations, and those are always ghost (at least in this example).